File: leetcode/problems/language-specific/go/125.valid-palindrome.go
/*
 * @lc app=leetcode id=125 lang=golang
 *
 * [125] Valid Palindrome
 */

// @lc code=start
func isPalindrome(s string) bool {
	if len(s) <= 1 {
		return true
	}

	var isAlphanumeric = regexp.MustCompile(`^[a-zA-Z0-9]$`).MatchString
	left, right := 0, len(s)-1
	for left < right {
		if !isAlphanumeric(string(s[left])) {
			left++
		} else if !isAlphanumeric(string(s[right])) {
			right--
		} else if strings.ToLower(string(s[left])) != strings.ToLower(string(s[right])) {
			return false
		} else {
			left++
			right--
		}

	}

	return true
}

// @lc code=end

File: leetcode/problems/problem-specific/125.valid-palindrome/125.valid-palindrome.py
#
# @lc app=leetcode id=125 lang=python3
#
# [125] Valid Palindrome
#

# @lc code=start

# * Two Pointers Solution | O(n) Time | O(1) Space
# * n -> The length of s string

class Solution:
    def isPalindrome(self, s: str) -> bool:
        s_length = len(s)
        if s_length <= 1:
            return True

        left, right = 0, s_length-1
        while left < right:
            if not s[left].isalnum():
                left += 1
            elif not s[right].isalnum():
                right -= 1
            elif s[left].lower() != s[right].lower():
                return False
            else:
                left += 1
                right -= 1

        return True

# @lc code=end
